### PR TITLE
Command line arguments

### PR DESCRIPTION
**Pull request by kofigumbs**
@trptcolin @robertjlooby Feature-wise, this adds various command line options (see README.md)

### CODE REVIEW

**Comment by trptcolin** [comment]
There's some weird behavior here where it asks for user input before printing out the available options:

```
colin:tictactoe-java/ (command-line-arguments) $ java -jar build/libs/tictactoe-java.jar --help
Welcome to Tic Tac Toe!

Would you like to go first? (y/n) y
Make a move by entering an empty space id
usage: java -jar tictactoe-java.jar [--size <number>] [--padding <number>] [--human|--minimax]
```

Also, this PR takes the code coverage down from 97% to 92% - let's get that back up above 95% (and to 100% if possible).

File: src/me/hkgumbs/tictactoe/main/java/Main.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            new FormatterConfiguration()
+    };
+
+    public static void main(String[] args) {
+        Simulation simulation = new DefaultSimulation();
+        List<String> arguments = new ArrayList<>(Arrays.asList(args));
+
+        try {
+            for (Configuration configuration : CONFIGURATIONS)
+                configuration.apply(arguments, simulation);
```
Comment (trptcolin) [inline src/me/hkgumbs/tictactoe/main/java/Main.java]: This design is nice, because it means you can add new configurations pretty easily.

It might be nice to separate this logic into another class, because looks like it'd be useful to test (things like catching exceptions, what to do if arguments are empty, how the configurations are applied, etc.). And `main()` is usually pretty tricky to get comprehensive tests around.

File: src/me/hkgumbs/tictactoe/main/java/simulation/SizeConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package me.hkgumbs.tictactoe.main.java.simulation;
+
+import java.util.List;
+
+public class SizeConfiguration implements Configuration {
+
+    private static final String KEY = "--size";
+    private static final int DEFAULT_SIZE = 3;
+
+    private int extractSize(List<String> args) {
```
Comment (trptcolin) [inline src/me/hkgumbs/tictactoe/main/java/simulation/SizeConfiguration.java]: This method looks pretty much identical to `extractPadding` in `FormatterConfiguration`. Surely there's a way to DRY that up - maybe a small helper class with (at least) one method that takes the argument list, the argument name, and the default value?

File: src/me/hkgumbs/tictactoe/test/java/simulation/RulesConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        OutputStream outputStream = new ByteArrayOutputStream();
+        Configuration[] configurations = new Configuration[]{
+                new SizeConfiguration(),
+                new RulesConfiguration(outputStream)
+        };
+        try {
+            for (Configuration configuration : configurations)
+                configuration.apply(Arrays.asList(), simulation);
+        } catch (Configuration.CannotApplyException e) {
+            e.printStackTrace();
```
Comment (trptcolin) [inline src/me/hkgumbs/tictactoe/test/java/simulation/RulesConfigurationTest.java]: What should happen in a test run if this exception gets thrown? Should the test fail? The way this is written, we'd see a stack trace, but no test would fail.

File: src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+
+    public Simulation configure(String... arguments) {
+        Simulation simulation = new DefaultSimulation();
+        List<String> args = new ArrayList<>(Arrays.asList(arguments));
+        Configuration configuration = new SizeConfiguration();
+        try {
+            configuration.apply(args, simulation);
+        } catch (Configuration.CannotApplyException e) {
+            e.printStackTrace();
```
Comment (trptcolin) [inline src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java]: Same question as before: should tests fail if an exception is caught here?

File: src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        Simulation simulation = new DefaultSimulation();
+        List<String> args = new ArrayList<>(Arrays.asList("--size", "zx", "5"));
+        Configuration configuration = new SizeConfiguration();
+        boolean flag = false;
+        try {
+            configuration.apply(args, simulation);
+        } catch (Configuration.CannotApplyException e) {
+            flag = true;
+        }
+        assertTrue(flag);
```
Comment (trptcolin) [inline src/me/hkgumbs/tictactoe/test/java/simulation/SizeConfigurationsTest.java]: It's always tricky to interpret error messages that say things like "Expected true, got false". I'd suggest adding a descriptive message using [this arity of `assertTrue`](http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertTrue(java.lang.String, boolean)) that allows a message to be displayed.

Another thing you could do to make this more readable would be to rename `flag` to something more descriptive: *why* does this thing get set to `true`?

**Comment by trptcolin** [comment]
Lots of good stuff in here. To name a few: some really nice renames, extracting the `Configuration` objects, and taking much better advantage of your `Player` abstraction.

File: src/me/hkgumbs/tictactoe/main/java/simulation/Extractor.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package me.hkgumbs.tictactoe.main.java.simulation;
+
+import java.util.List;
+
+public class Extractor {
+    public static int extract(List<String> args, String key, int defaultValue)
```
Comment (robertjlooby) [inline src/me/hkgumbs/tictactoe/main/java/simulation/Extractor.java]: `Extractor.extract` isn't a very descriptive name...

**Comment by hkgumbs** [comment]
@trptcolin @robertjlooby I am concerned about the Simulation interface, since it basically just exposes the private variables in DefaultSimulation. I read in Clean Code about data structures, and just wanted to make sure it was still a good practice since it was new to me. I'm thinking of just having a Simulation class that just has the 5 or so public variables that keep track of the game, then moving the state change logic to a separate controller class. Then I can pass the Simulation around to the Players as well instead of the setter/getter exposing I'm doing there.

File: src/me/hkgumbs/tictactoe/main/java/Main.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class Main {
+    private static final String USAGE = "usage: java -jar tictactoe-java.jar" +
+            " [--size <number>]" +
+            " [--padding <number>]" +
+            " [--humans|--minimax]";
```
Comment (robertjlooby) [inline src/me/hkgumbs/tictactoe/main/java/Main.java]: It would be nice to document the options in the README

**Comment by trptcolin** [comment]
@hkgumbs I agree: by providing a wide interface like `Simulation`, you make it possible for Players to do things like swap out the other player (!) - what if a Player implementation wanted to cheat and pass in a TerriblePlayer implementation? 

So I agree that this thing might have 2 jobs: wiring and playing the game.

One nice thing about the `Simulation` interface, even though it looks like it's just exposing private variables, is that it decouples the storage mechanism from how you tell the thing to get/change state. I think you can retain this benefit with the split you're suggesting in your comment.

File: src/me/hkgumbs/tictactoe/main/java/Main.java (hunk 7 above)
Comment (hkgumbs) [inline src/me/hkgumbs/tictactoe/main/java/Main.java]: @robertjlooby done

**Comment by hkgumbs** [comment]
@trptcolin @robertjlooby I apologize for yet another lengthy commit. As I was pulling DefaultSimulation apart, I had to refactor many times, and it seemed like any intermediate state would have left the game broken

**Comment by hkgumbs** [comment]
@trptcolin @robertjlooby 
- [X] Speed up Minimax to less than 1.5 seconds
- [X] Removed double printing/asking in human v human

File: src/me/hkgumbs/tictactoe/main/java/rules/DefaultRules.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            Set<Integer> marks = board.getSpaceIds(mark);
+            for (Set combination : winningCombinations)
+                if (marks.containsAll(combination))
+                    return mark;
+        }
+        return null;
+    }
+
+    @Override
+    public void printWinnerMessage(Board board) {
```
Comment (hkgumbs) [inline src/me/hkgumbs/tictactoe/main/java/rules/DefaultRules.java]: Not sure if this is the right decision here, since it forces the Rules to deal with output, but it would be duplicated if I put it in Player concretions, and it doesn't seem like a big enough idea fro its own class.

Comment (robertjlooby) [inline src/me/hkgumbs/tictactoe/main/java/rules/DefaultRules.java]: I agree the `Rules` shouldn't care about output. This and `generateWinnerMessage` should be moved somewhere else so the `Rules` don't care about actual output (i.e. the `Rules` class should not depend on `OutputStream` at all)

**Comment by hkgumbs** [comment]
@robertjlooby @trptcolin iterated on all of the feedback here. I think this and #4 finish out this project.

**Comment by trptcolin** [comment]
@hkgumbs looks good to me. Very nice speed improvement. It took awhile for me to figure out how to play human vs. minimax (user must omit flags), so it might be nice to provide some default. I bet you've got that solved w/ the dumb computer branch though, so I'll go ahead & merge.
